### ISSUE TITLE
Need a standardized copyright message in every source file

### ISSUE THREAD

**Issue by sherm1** [RobotLocomotion/drake#1805]
Each file in an open source project should start with a standardized copyright/license block. Users can typically do anything with the code _except_ remove the block.

We need to decide:
- what license to use (I favor Apache 2.0)
- who should hold the copyright (that determines who must agree on a change to the license)
- what other information (if any) should be included (e.g. credit to funding agencies & contributors)
- how the header should be formatted

and then figure out how to update all the existing files.

Here are a couple of examples to get started: 
- [Google tensorflow](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/graph/equal_graph_def.h#L1-L14) (minimalist: no credit or copyright holding for anyone but Google); OSRF uses this format also
- [Simbody](https://github.com/simbody/simbody/blob/master/SimTKcommon/Mechanics/include/SimTKcommon/internal/Rotation.h#L4-L25) (flashy: credits funding agency, provides authors copyright control, credits contributors)

**Comment by RussTedrake** [comment]
thanks!  i would err on the side of the minimal.

**Comment by liangfok** [comment]
We can use a bash script to update all existing source files in a directory with the copyright text: 

http://unix.stackexchange.com/questions/125421/how-to-prepend-a-particular-text-file-contents-to-every-text-file-in-a-directory

Ideally, the git commit should be done by a robot to make it clear the change was automated.

**Comment by jamiesnape** [comment]
Resolving this would keep cpplint happy too.

**Comment by jamiesnape** [comment]
I would like to fix the hundreds of lint errors due to this, so I propose the following.

C++ / Java:
```
// Copyright 2016 Robot Locomotion Group @ CSAIL. All rights reserved.
//
// All components of Drake are licensed under the BSD 3-Clause License.
// See LICENSE.TXT or http://drake.mit.edu/ for details.
```

CMake / Python / Shell:
```
# Copyright 2016 Robot Locomotion Group @ CSAIL. All rights reserved.
#
# All components of Drake are licensed under the BSD 3-Clause License.
# See LICENSE.TXT or http://drake.mit.edu/ for details.
```

MATLAB:
```
% Copyright 2016 Robot Locomotion Group @ CSAIL. All rights reserved.
%
% All components of Drake are licensed under the BSD 3-Clause License.
% See LICENSE.TXT or http://drake.mit.edu/ for details.
```
The year is the year the file was created. The template that the linter follows is `Copyright YEAR ORGANIZATION`.

**Comment by RussTedrake** [comment]
@jamiesnape -- we've decided to step back and resolve the copyright correctly before deciding.  And I believe you can disable the errors with a one line argument to the lint exe.  @david-german-tri can you remind me what that was?

**Comment by jamiesnape** [comment]
Sure, I can can disable it, if you prefer. I know the command.

**Comment by sherm1** [comment]
> we've decided to step back and resolve the copyright correctly before deciding

Just so it won't get forgotten -- both MIT and TRI are likely to be the copyright holders.

**Comment by david-german-tri** [comment]
For anyone else following along, the flag Russ mentioned is

`cpplint --filter="-legal/copyright"`

**Comment by RussTedrake** [comment]
offline discussion is in progress.

**Comment by jwnimmer-tri** [comment]
No, we still need to add some TBD boilerplate atop every file.  It's still on my TODO list.  At this point, it is more difficult because will probably need to `git blame` our way through the list of copyright holders for each file.

**Comment by jwnimmer-tri** [comment]
https://softwarefreedom.org/resources/2012/ManagingCopyrightInformation.html has some good advice.

**Comment by jamiesnape** [comment]
I think I like their suggestions on centralized licenses and their suggested boilerplate more or less (though I would not be bothered if we skipped the boilerplate). I think individual attribution in files is probably barely viable (or worth the effort) now, so if we need (corporate/institutional) names for copyright purposes beyond source control, an overall `AUTHORS` file is going to be easiest (see Bazel for one example, of course).

I would also like to see the `LICENSE` file be a verbatim dump of the BSD license. We are mixing legal text with advisory, non-legal text at the moment. If we need notes on exceptions, they should be in `third_party`, `tools/workspace`, in line in the relevant files, or in a `NOTICES file`, I think.

**Comment by jamiesnape** [comment]
Since I have been running automated license checking tools through our code and our dependencies code the last few days, I would just say that having something, anything, is important otherwise it needs manual inspection, which is not practical in a project the size of Drake, especially as we have third-party code in tree.

At a minimum you want on one line the word `Copyright`, the year of creation, and some entity even if it is "The Drake Development Team". Assuming we stick with BSD 3-clause, on a different line, you would want the words "BSD 3-clause license", or "BSD-3-Clause", or "BSD-3-clause". These should be the first lines in the file, modulo shebangs or codelines Whatever is chosen, running the Debain license checker through the tree and checking the license and copyright fields are identified correctly is a good idea.

**Comment by sherm1** [comment]
This issue is 10 years old now. I think it's safe to say we're not going to do it. Closing.

**Comment by jwnimmer-tri** [comment]
I still plan to do it.

**Comment by sherm1** [comment]
Why? The files have been on the loose for 10 years with no per-file copyright message. If there really is value to a per-file message, adding it now won't change anything about the pre-existing ~1 million lines that are already available without it. No harm if you really want to do it, but I think the cat's out of the bag already.

**Comment by jwnimmer-tri** [comment]
The `SPDX-License-Identifier:` clause (which is the solution here) is an easy way to explain what license each file in our codebase is available under.  It helps us as maintainers know the terms for each file (not all of which are `BSD-3-Clause`).  Even aside from questions of other people copying or using our code, it still helps *us* understand what's in our own house.

Similarly, it helps other people who want to borrow parts of our code understand what their options and obligations are.

The point is clear communication.  The extra comment doesn't _change_ the copyrighted nature of the files (nothing is "out of the bag").  It only more clearly explains the state of affairs, and in a way that travels in-band with the files as people fork them (e.g., https://github.com/wep21/cmake_configure_file/blob/main/cmake_configure_file.py or https://github.com/ApolloAuto/apollo/blob/master/tools/install/install.bzl).
